source/utilities/server.ts

```diff
@@ -0,0 +1,126 @@
+// source/utilities/server.ts
+// Run the server with the given configuration.
+
+import http from 'node:http';
+import https from 'node:https';
+import { readFile } from 'node:fs/promises';
+import handler from 'serve-handler';
+import compression from 'compression';
+import isPortReachable from 'is-port-reachable';
+import { getNetworkAddress, registerCloseListener } from './http.js';
+import { promisify } from './promise.js';
+import type { IncomingMessage, ServerResponse } from 'node:http';
+import type { AddressInfo } from 'node:net';
+import type {
+  Configuration,
+  Options,
+  ParsedEndpoint,
+  Port,
+  ServerAddress,
+} from '../types.js';
+
+const compress = promisify(compression());
+
+/**
+ * Starts the server and makes it listen on the given endpoint.
+ *
+ * @param endpoint - The endpoint to listen on.
+ * @param config - The configuration for the `serve-handler` middleware.
+ * @param args - The arguments passed to the CLI.
+ * @returns The address of the server.
+ */
+export const startServer = async (
+  endpoint: ParsedEndpoint,
+  config: Partial<Configuration>,
+  args: Options,
+  previous?: Port,
+): Promise<ServerAddress> => {
+  // Define the request handler for the server.
+  const serverHandler = async (
+    request: IncomingMessage,
+    response: ServerResponse,
+  ): Promise<void> => {
+    if (args['--cors']) response.setHeader('Access-Control-Allow-Origin', '*');
+    // @ts-expect-error The `compression` library uses the Express Request type
+    // instead of the native HTTP IncomingMessage type.
+    if (!args['--no-compression']) await compress(request, response);
+
+    // Let the `serve-handler` module do the rest.
+    await handler(request, response, config);
+  };
+
+  // Create the server.
+  const httpMode = args['--ssl-cert'] && args['--ssl-key'] ? 'https' : 'http';
+  const sslPass = args['--ssl-pass'];
+  const server =
+    httpMode === 'https'
+      ? https.createServer(
+          {
+            key: await readFile(args['--ssl-key']),
+            cert: await readFile(args['--ssl-cert']),
+            passphrase: sslPass ? await readFile(sslPass, 'utf8') : '',
+          },
+          serverHandler, // eslint-disable-line @typescript-eslint/no-misused-promises
+        )
+      : http.createServer(serverHandler); // eslint-disable-line @typescript-eslint/no-misused-promises
```

Comment:
You could avoid this disable by not making `serverHandler` async, and adding a `catch` to the awaited `handler`.

Reply:
Ahh okay, will do!

Reply:
I wrapped the contents of the `serverHandler` function in an async function instead, as the `compress` call was also async.